In [71]:
import psycopg2
from psycopg2 import OperationalError

In [82]:
def create_connection():
    """Estabelece uma conexão com o banco de dados PostgreSQL."""
    try:
        connection = psycopg2.connect(
            user="postgres",
            password="admin",
            host="localhost",
            port="5432",
            database="big_data"
        )
        print("Conexão com o banco de dados estabelecida com sucesso!")
        return connection

    except OperationalError as e:
        print(f"Erro ao conectar ao banco de dados: {e}")
        return None

def close_connection(connection):
    """Fecha a conexão com o banco de dados."""
    if connection:
        connection.close()
        print("Conexão com o banco de dados fechada.")

In [115]:
# Criar uma conexão
conn = create_connection()

# Criar um cursor
cursor = conn.cursor()

Conexão com o banco de dados estabelecida com sucesso!


In [114]:
# encerrar conexão
close_connection(conn)

Conexão com o banco de dados fechada.


In [63]:
cursor.execute('''CREATE TABLE EMPREGADO (
  ssn char(9) NOT NULL,
  pnome varchar(30) NOT NULL,
  inicialm char(1) default NULL,
  unome varchar(30) NOT NULL,
  sexo char(1) NOT NULL,
  endereco varchar(50) default NULL,
  datanasc date default NULL,
  superssn char(9) default NULL,
  dno char(1) NOT NULL,
  salario float default NULL,
  PRIMARY KEY  (ssn)
) ;
''')

In [64]:
cursor.execute("INSERT INTO EMPREGADO VALUES ('123456789', 'Joao', 'B', 'Souza', 'M', 'R. 24 de maio, 1500 - Curitiba - PR', '1965-01-09', '333445555', '5', '30000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('333445555', 'Fabio', 'T', 'Will', 'M', 'R. Alagoas, 325 - Curitiba - PR', '1955-12-08', '888665555', '5', '40000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('999887777', 'Alice', 'J', 'Zebra', 'F', 'R. Ribeira, 98 - Pinhais - PR', '1968-01-19', '987654321', '4', '25000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('987654321', 'Jennifer', 'S', 'Wallace', 'F', 'R. Berry, 291 - Colombo - PR', '1941-06-20', '888665555', '4', '43000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('666884444', 'Ricardo', 'K', 'Nantes', 'M', 'Av. Figueira, 55 - Almirante Tamandare - PR', '1962-09-15', '333445555', '5', '38000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('453453453', 'Jussara', 'A', 'Pereira', 'F', 'Trav. da Lapa, 23 - Curitiba - PR', '1972-07-31', '333445555', '5', '25000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('987987987', 'Alberto', 'V', 'Medeiros', 'M', 'R. Ceará, 1245 - Curitiba - PR', '1969-03-29', '987654321', '4', '25000.00');")
cursor.execute("INSERT INTO EMPREGADO VALUES ('888665555', 'Joaquim', 'E', 'Brito', 'M', 'R. Stone, 450 - Curitiba - PR', '1937-11-10', null, '1', '55000.00');")

In [65]:
cursor.execute('''CREATE TABLE DEPARTAMENTO (
  dnumero char(1) NOT NULL,
  dnome varchar(30) NOT NULL,
  gerssn char(9) NOT NULL,
  gerdatainicio date NOT NULL,
  PRIMARY KEY  (dnumero)
) ;
''')

In [66]:
cursor.execute("INSERT INTO DEPARTAMENTO VALUES ('5', 'Pesquisa', '333445555', '1988-05-22');")
cursor.execute("INSERT INTO DEPARTAMENTO VALUES ('4', 'Administracao', '987654321', '1995-01-01');")
cursor.execute("INSERT INTO DEPARTAMENTO VALUES ('1', 'Sede Administrativa', '888665555', '1981-06-19');")

In [67]:
cursor.execute('''
CREATE TABLE DEPENDENTE (
  essn char(9) NOT NULL,
  nomedep varchar(30) NOT NULL default '',
  sexodep char(1) NOT NULL,
  datanascdep date NOT NULL,
  parentesco varchar(30) NOT NULL,
  PRIMARY KEY  (essn,nomedep)
) ;
''')

In [68]:
cursor.execute("INSERT INTO DEPENDENTE VALUES ('333445555', 'Alice', 'F', '1986-04-05', 'FILHA');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('333445555', 'Teodoro', 'M', '1983-10-25', 'FILHO');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('333445555', 'Joana', 'F', '1958-05-03', 'CONJUGE');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('987654321', 'Abdala', 'M', '1942-02-28', 'CONJUGE');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('123456789', 'Michel', 'M', '1988-01-04', 'FILHO');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('123456789', 'Alice', 'F', '1988-12-30', 'FILHA');")
cursor.execute("INSERT INTO DEPENDENTE VALUES ('123456789', 'Elizabete', 'F', '1967-05-05', 'CONJUGE');")

In [69]:
cursor.execute('''
CREATE TABLE LOCALIZACAO (
  dnum char(1) NOT NULL,
  dlocalizacao varchar(30) NOT NULL,
  PRIMARY KEY  (dnum,dlocalizacao)
) ;
''')

In [70]:
cursor.execute("INSERT INTO LOCALIZACAO VALUES ('1', 'Curitiba');")
cursor.execute("INSERT INTO LOCALIZACAO VALUES ('4', 'Colombo');")
cursor.execute("INSERT INTO LOCALIZACAO VALUES ('5', 'Araucaria');")
cursor.execute("INSERT INTO LOCALIZACAO VALUES ('5', 'Curitiba');")
cursor.execute("INSERT INTO LOCALIZACAO VALUES ('5', 'Pinhais');")

In [71]:
cursor.execute('''
CREATE TABLE PROJETO (
  pnumero char(2) NOT NULL,
  pjnome varchar(30) NOT NULL,
  plocal varchar(30) NOT NULL,
  dnum char(1) NOT NULL,
  PRIMARY KEY  (pnumero)
) ;
''')

In [72]:
cursor.execute("INSERT INTO PROJETO VALUES ('1', 'ProdutoX', 'Pinhais', '5');")
cursor.execute("INSERT INTO PROJETO VALUES ('2', 'ProdutoY', 'Araucaria', '5');")
cursor.execute("INSERT INTO PROJETO VALUES ('3', 'ProdutoZ', 'Curitiba', '5');")
cursor.execute("INSERT INTO PROJETO VALUES ('10', 'Automatizacao', 'Pinhais', '4');")
cursor.execute("INSERT INTO PROJETO VALUES ('20', 'Reorganizacao', 'Curitiba', '1');")
cursor.execute("INSERT INTO PROJETO VALUES ('30', 'Novos Beneficios', 'Pinhais', '4');")

In [73]:
cursor.execute('''
CREATE TABLE TRABALHA (
  essn char(9) NOT NULL,
  pno char(2) NOT NULL,
  horas float default NULL,
  PRIMARY KEY  (essn,pno)
) ;
''')

In [74]:
cursor.execute("INSERT INTO TRABALHA VALUES ('123456789', '1', '32.5');")
cursor.execute("INSERT INTO TRABALHA VALUES ('123456789', '2', '7.5');")
cursor.execute("INSERT INTO TRABALHA VALUES ('666884444', '3', '40');")
cursor.execute("INSERT INTO TRABALHA VALUES ('453453453', '1', '20');")
cursor.execute("INSERT INTO TRABALHA VALUES ('453453453', '2', '20');")
cursor.execute("INSERT INTO TRABALHA VALUES ('333445555', '2', '10');")
cursor.execute("INSERT INTO TRABALHA VALUES ('333445555', '3', '10');")
cursor.execute("INSERT INTO TRABALHA VALUES ('333445555', '10', '10');")
cursor.execute("INSERT INTO TRABALHA VALUES ('333445555', '20', '10');")
cursor.execute("INSERT INTO TRABALHA VALUES ('999887777', '30', '30');")
cursor.execute("INSERT INTO TRABALHA VALUES ('999887777', '10', '10');")
cursor.execute("INSERT INTO TRABALHA VALUES ('987987987', '10', '35');")
cursor.execute("INSERT INTO TRABALHA VALUES ('987987987', '30', '5');")
cursor.execute("INSERT INTO TRABALHA VALUES ('987654321', '30', '20');")
cursor.execute("INSERT INTO TRABALHA VALUES ('987654321', '20', '15');")
cursor.execute("INSERT INTO TRABALHA VALUES ('888665555', '20', null);")

In [75]:
cursor.execute("ALTER TABLE EMPREGADO ADD CONSTRAINT fk_empregado_empregado  FOREIGN KEY (superssn) REFERENCES  EMPREGADO (ssn);")
cursor.execute("ALTER TABLE EMPREGADO ADD CONSTRAINT fk_empregado_departamento FOREIGN KEY  (dno) REFERENCES  DEPARTAMENTO (dnumero) ;")
cursor.execute("ALTER TABLE DEPARTAMENTO ADD CONSTRAINT fk_departamento_empregado FOREIGN KEY (gerssn) REFERENCES  EMPREGADO (ssn) ;")
cursor.execute("ALTER TABLE DEPENDENTE ADD CONSTRAINT fk_dependente_empregado FOREIGN KEY (essn) REFERENCES EMPREGADO (ssn) ;")
cursor.execute("ALTER TABLE LOCALIZACAO ADD CONSTRAINT fk_localizacao_departamento FOREIGN KEY (dnum) REFERENCES  DEPARTAMENTO (dnumero) ;")
cursor.execute("ALTER TABLE PROJETO ADD CONSTRAINT fk_projeto_departamento FOREIGN KEY (dnum) REFERENCES  DEPARTAMENTO (dnumero) ;")
cursor.execute("ALTER TABLE TRABALHA ADD CONSTRAINT fk_trabalha_empregado FOREIGN KEY (essn) REFERENCES  EMPREGADO (ssn) ;")
cursor.execute("ALTER TABLE TRABALHA ADD CONSTRAINT fk_trabalha_projeto FOREIGN KEY (pno) REFERENCES  PROJETO (pnumero) ;")
               
# Commitar as transações
conn.commit()

a) Insira >‘943775543’, ‘Roberto’, ‘F’, ‘Silva’, ‘M’, ‘Rua X, 22 – Araucária – PR’, ‘1952-06-21’,
‘888665555’, ‘1’, 58000< em EMPREGADO.

In [76]:
cursor.execute("INSERT INTO EMPREGADO VALUES ('943775543', 'Roberto', 'F', 'Silva', 'M', 'Rua X, 22 – Araucária – PR', '1952-06-21', '888665555', '1', '58000');")
# Confirmar a transação
conn.commit()

In [77]:
cursor.execute("SELECT * FROM EMPREGADO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', 'Joao', 'B', 'Souza', 'M', 'R. 24 de maio, 1500 - Curitiba - PR', datetime.date(1965, 1, 9), '333445555', '5', 30000.0)
('333445555', 'Fabio', 'T', 'Will', 'M', 'R. Alagoas, 325 - Curitiba - PR', datetime.date(1955, 12, 8), '888665555', '5', 40000.0)
('999887777', 'Alice', 'J', 'Zebra', 'F', 'R. Ribeira, 98 - Pinhais - PR', datetime.date(1968, 1, 19), '987654321', '4', 25000.0)
('987654321', 'Jennifer', 'S', 'Wallace', 'F', 'R. Berry, 291 - Colombo - PR', datetime.date(1941, 6, 20), '888665555', '4', 43000.0)
('666884444', 'Ricardo', 'K', 'Nantes', 'M', 'Av. Figueira, 55 - Almirante Tamandare - PR', datetime.date(1962, 9, 15), '333445555', '5', 38000.0)
('453453453', 'Jussara', 'A', 'Pereira', 'F', 'Trav. da Lapa, 23 - Curitiba - PR', datetime.date(1972, 7, 31), '333445555', '5', 25000.0)
('987987987', 'Alberto', 'V', 'Medeiros', 'M', 'R. Ceará, 1245 - Curitiba - PR', datetime.date(1969, 3, 29), '987654321', '4', 25000.0)
('888665555', 'Joaquim', 'E', 'Brito', 'M', 'R. St

b) Insira >‘4’, ProdutoA’, ‘Araucaria’, ‘2’< em PROJETO

In [41]:
cursor.execute("INSERT INTO PROJETO VALUES ('4', 'ProdutoA', 'Araucaria', '2');")

ForeignKeyViolation: ERRO:  inserção ou atualização em tabela "projeto" viola restrição de chave estrangeira "fk_projeto_departamento"
DETAIL:  Chave (dnum)=(2) não está presente na tabela "departamento".


In [78]:
cursor.execute("SELECT * FROM DEPARTAMENTO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('5', 'Pesquisa', '333445555', datetime.date(1988, 5, 22))
('4', 'Administracao', '987654321', datetime.date(1995, 1, 1))
('1', 'Sede Administrativa', '888665555', datetime.date(1981, 6, 19))


A tabela departamento não possui nenhum registro dnumero igual a 2, e esse atributo é chave primária, portanto ao tentar inserir um novo registro em projeto, com dnumero igual a 2, o banco de dados irá retorna um erro de chave estrangeira, pois não existe um departamento com dnumero igual a 2.

c) Insira >‘4’, ‘Produção’, ‘943775543’, ‘1998-10-01’< em DEPARTAMENTO.

In [50]:
cursor.execute("INSERT INTO DEPARTAMENTO VALUES ('4', 'Produção', '943775543', '1998-10-01');")
# Confirmar a transação
conn.commit()

UniqueViolation: ERRO:  duplicar valor da chave viola a restrição de unicidade "departamento_pkey"
DETAIL:  Chave (dnumero)=(4) já existe.


dnumero é chave primária na tabela departamento, e o valor 4 já existe, portanto ao tentar inserir um novo registro em departamento, com dnumero igual a 4, o banco de dados irá retorna um erro de chave primária, pois já existe um departamento com dnumero igual a 4.

d) Insira >‘677678989’, null, 40.0< em TRABALHA

In [54]:
cursor.execute("INSERT INTO TRABALHA VALUES ('677678989', 'null', '40.0');")
# Confirmar a transação
conn.commit()

StringDataRightTruncation: ERRO:  valor é muito longo para tipo character(2)


O campo pno aceita apenas char com 2 caracteres, e o valor inserido 'null' possui 4 caracteres, portanto o banco de dados irá retornar um erro de tipo de dados.

e) Insira >‘453453453’, ‘Joao’, ‘M’, ‘1970-12-12’, ‘CONJUGE’< em DEPENDENTE.

In [79]:
cursor.execute("INSERT INTO DEPENDENTE VALUES ('453453453', 'Joao', 'M', '1970-12-12', 'CONJUGE');")
# Confirmar a transação
conn.commit()

In [80]:
cursor.execute("SELECT * FROM DEPENDENTE;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('333445555', 'Alice', 'F', datetime.date(1986, 4, 5), 'FILHA')
('333445555', 'Teodoro', 'M', datetime.date(1983, 10, 25), 'FILHO')
('333445555', 'Joana', 'F', datetime.date(1958, 5, 3), 'CONJUGE')
('987654321', 'Abdala', 'M', datetime.date(1942, 2, 28), 'CONJUGE')
('123456789', 'Michel', 'M', datetime.date(1988, 1, 4), 'FILHO')
('123456789', 'Alice', 'F', datetime.date(1988, 12, 30), 'FILHA')
('123456789', 'Elizabete', 'F', datetime.date(1967, 5, 5), 'CONJUGE')
('453453453', 'Joao', 'M', datetime.date(1970, 12, 12), 'CONJUGE')


f) Remova as tuplas de TRABALHA com ESSN = ‘333445555’

In [81]:
cursor.execute("DELETE FROM TRABALHA WHERE essn = '333445555';")
# Confirmar a transação
conn.commit()

In [82]:
cursor.execute("SELECT * FROM TRABALHA;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', '1 ', 32.5)
('123456789', '2 ', 7.5)
('666884444', '3 ', 40.0)
('453453453', '1 ', 20.0)
('453453453', '2 ', 20.0)
('999887777', '30', 30.0)
('999887777', '10', 10.0)
('987987987', '10', 35.0)
('987987987', '30', 5.0)
('987654321', '30', 20.0)
('987654321', '20', 15.0)
('888665555', '20', None)


g) Remova a tupla de EMPREGADO com SSN = ‘987654321’.

In [61]:
cursor.execute("DELETE FROM EMPREGADO WHERE ssn = '987654321';")
# Confirmar a transação
conn.commit()

ForeignKeyViolation: ERRO:  atualização ou exclusão em tabela "empregado" viola restrição de chave estrangeira "fk_empregado_empregado" em "empregado"
DETAIL:  Chave (ssn)=(987654321) ainda é referenciada pela tabela "empregado".


In [83]:
cursor.execute("SELECT * FROM EMPREGADO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', 'Joao', 'B', 'Souza', 'M', 'R. 24 de maio, 1500 - Curitiba - PR', datetime.date(1965, 1, 9), '333445555', '5', 30000.0)
('333445555', 'Fabio', 'T', 'Will', 'M', 'R. Alagoas, 325 - Curitiba - PR', datetime.date(1955, 12, 8), '888665555', '5', 40000.0)
('999887777', 'Alice', 'J', 'Zebra', 'F', 'R. Ribeira, 98 - Pinhais - PR', datetime.date(1968, 1, 19), '987654321', '4', 25000.0)
('987654321', 'Jennifer', 'S', 'Wallace', 'F', 'R. Berry, 291 - Colombo - PR', datetime.date(1941, 6, 20), '888665555', '4', 43000.0)
('666884444', 'Ricardo', 'K', 'Nantes', 'M', 'Av. Figueira, 55 - Almirante Tamandare - PR', datetime.date(1962, 9, 15), '333445555', '5', 38000.0)
('453453453', 'Jussara', 'A', 'Pereira', 'F', 'Trav. da Lapa, 23 - Curitiba - PR', datetime.date(1972, 7, 31), '333445555', '5', 25000.0)
('987987987', 'Alberto', 'V', 'Medeiros', 'M', 'R. Ceará, 1245 - Curitiba - PR', datetime.date(1969, 3, 29), '987654321', '4', 25000.0)
('888665555', 'Joaquim', 'E', 'Brito', 'M', 'R. St

O atributo ssn da tabela empregado, é chave primária e estrangeira, sendo referenciada por superssn na mesma tabela, e por essn na tabela trabalha, portanto ao tentar remover a tupla de empregado com ssn igual a 987654321, o banco de dados irá retornar um erro de chave estrangeira, pois existem registros na mesma tabela e em outras tabelas que referenciam essa tupla.

h) Remova a tupla de PROJETO com PJNOME = ‘ProdutoX’

In [67]:
cursor.execute("DELETE FROM PROJETO WHERE PJNOME = 'ProdutoX';")
# Confirmar a transação
conn.commit()

ForeignKeyViolation: ERRO:  atualização ou exclusão em tabela "projeto" viola restrição de chave estrangeira "fk_trabalha_projeto" em "trabalha"
DETAIL:  Chave (pnumero)=(1 ) ainda é referenciada pela tabela "trabalha".


In [84]:
cursor.execute("SELECT * FROM PROJETO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('1 ', 'ProdutoX', 'Pinhais', '5')
('2 ', 'ProdutoY', 'Araucaria', '5')
('3 ', 'ProdutoZ', 'Curitiba', '5')
('10', 'Automatizacao', 'Pinhais', '4')
('20', 'Reorganizacao', 'Curitiba', '1')
('30', 'Novos Beneficios', 'Pinhais', '4')


In [85]:
cursor.execute("SELECT * FROM TRABALHA;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', '1 ', 32.5)
('123456789', '2 ', 7.5)
('666884444', '3 ', 40.0)
('453453453', '1 ', 20.0)
('453453453', '2 ', 20.0)
('999887777', '30', 30.0)
('999887777', '10', 10.0)
('987987987', '10', 35.0)
('987987987', '30', 5.0)
('987654321', '30', 20.0)
('987654321', '20', 15.0)
('888665555', '20', None)


Pnumero é chave primária em projeto, e chave estrangeira em trabalha, portanto ao tentar remover a tupla de projeto com pjnome igual a ProdutoX, o banco de dados irá retornar um erro de chave estrangeira, pois existem registros na tabela trabalha que referenciam essa tupla.

i) Modifique o GERSSN e GERDATAINICIO da tupla de DEPARTAMENTO com DNUMERO = 5
para ‘123456789’ e ‘1999-01-10’, respectivamente.

In [72]:
cursor.execute("UPDATE DEPARTAMENTO SET GERSSN = '123456789' , GERDATAINICIO ='1999-01-10'  WHERE DNUMERO = '5';")
# Confirmar a transação
conn.commit()

In [73]:
cursor.execute("SELECT * FROM DEPARTAMENTO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('4', 'Administracao', '987654321', datetime.date(1995, 1, 1))
('1', 'Sede Administrativa', '888665555', datetime.date(1981, 6, 19))
('5', 'Pesquisa', '123456789', datetime.date(1999, 1, 10))


j) Modifique o atributo SUPERSSN da tupla EMPREGADO com SSN = ‘999887777’ para
‘943775543’.

In [74]:
cursor.execute("UPDATE EMPREGADO SET SUPERSSN = '943775543' WHERE SSN = '999887777';")
# Confirmar a transação
conn.commit()

In [75]:
cursor.execute("SELECT * FROM EMPREGADO;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', 'Joao', 'B', 'Souza', 'M', 'R. 24 de maio, 1500 - Curitiba - PR', datetime.date(1965, 1, 9), '333445555', '5', 30000.0)
('333445555', 'Fabio', 'T', 'Will', 'M', 'R. Alagoas, 325 - Curitiba - PR', datetime.date(1955, 12, 8), '888665555', '5', 40000.0)
('987654321', 'Jennifer', 'S', 'Wallace', 'F', 'R. Berry, 291 - Colombo - PR', datetime.date(1941, 6, 20), '888665555', '4', 43000.0)
('666884444', 'Ricardo', 'K', 'Nantes', 'M', 'Av. Figueira, 55 - Almirante Tamandare - PR', datetime.date(1962, 9, 15), '333445555', '5', 38000.0)
('453453453', 'Jussara', 'A', 'Pereira', 'F', 'Trav. da Lapa, 23 - Curitiba - PR', datetime.date(1972, 7, 31), '333445555', '5', 25000.0)
('987987987', 'Alberto', 'V', 'Medeiros', 'M', 'R. Ceará, 1245 - Curitiba - PR', datetime.date(1969, 3, 29), '987654321', '4', 25000.0)
('888665555', 'Joaquim', 'E', 'Brito', 'M', 'R. Stone, 450 - Curitiba - PR', datetime.date(1937, 11, 10), None, '1', 55000.0)
('943775543', 'Roberto', 'F', 'Silva', 'M', 'Rua X, 22

l) Modifique o atributo HORAS da tupla de TRABALHA com ESSN = ‘999887777’ e PNO = ‘10’
para 5,0.

In [76]:
cursor.execute("UPDATE TRABALHA SET HORAS = '5' WHERE ESSN = '999887777' AND PNO = '10';")
# Confirmar a transação
conn.commit()

In [14]:
cursor.execute("SELECT * FROM TRABALHA;")
rows = cursor.fetchall()

# Iterar sobre os resultados e imprimir
for row in rows:
    print(row)

('123456789', '1 ', 32.5)
('123456789', '2 ', 7.5)
('666884444', '3 ', 40.0)
('453453453', '1 ', 20.0)
('453453453', '2 ', 20.0)
('999887777', '30', 30.0)
('987987987', '10', 35.0)
('987987987', '30', 5.0)
('987654321', '30', 20.0)
('987654321', '20', 15.0)
('888665555', '20', None)
('999887777', '10', 5.0)


## Atividade 2) Comandos DML: 
Considerando o banco de dados "empresa” apresentado na Atividade 1, escreva o comando DML adequado que:

a) Mostre o número de segurança social do empregado, o nome do dependente e o
parentesco, ordenado por ssn ascendente e parentesco descendente.

In [15]:
cursor.execute("SELECT e.ssn, d.nomedep, d.parentesco FROM EMPREGADO e JOIN DEPENDENTE d ON e.ssn = d.essn ORDER BY e.ssn ASC, d.parentesco DESC;")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('123456789', 'Michel', 'FILHO')
('123456789', 'Alice', 'FILHA')
('123456789', 'Elizabete', 'CONJUGE')
('333445555', 'Teodoro', 'FILHO')
('333445555', 'Alice', 'FILHA')
('333445555', 'Joana', 'CONJUGE')
('453453453', 'Joao', 'CONJUGE')
('987654321', 'Abdala', 'CONJUGE')


### **SELECT Clause**
```sql
SELECT e.ssn, d.nomedep, d.parentesco
```
- **SELECT** é usado para especificar as colunas que serão retornadas pelo query. Neste caso:
  - `e.ssn`: refere-se à coluna `ssn` da tabela `EMPREGADO`, que foi renomeada para `e` para facilitar as referências dentro da consulta.
  - `d.nomedep`: refere-se à coluna `nomedep` da tabela `DEPENDENTE`, que foi renomeada para `d`.
  - `d.parentesco`: refere-se à coluna `parentesco` da tabela `DEPENDENTE`.

### **FROM Clause**
```sql
FROM EMPREGADO e
```
- **FROM** indica a tabela principal de onde os dados serão selecionados. Aqui, a tabela `EMPREGADO` é renomeada como `e`. Essa renomeação é conhecida como aliasing e é muito útil para simplificar as referências à tabela em consultas mais complexas ou quando o mesmo nome de tabela é usado em múltiplas junções.

### **JOIN Clause**
```sql
JOIN DEPENDENTE d
```
- **JOIN** é utilizado para combinar linhas de duas ou mais tabelas baseadas em uma coluna relacionada entre elas. Aqui, a tabela `DEPENDENTE` é renomeada como `d`. Não é especificado o tipo de JOIN, o que implica que é um **INNER JOIN** por padrão. O INNER JOIN só retorna linhas que têm correspondência em ambas as tabelas.

### **ON Clause**
```sql
ON e.ssn = d.essn
```
- **ON** especifica a condição de junção. Neste caso, a consulta junta as tabelas `EMPREGADO` e `DEPENDENTE` onde o `ssn` da tabela `EMPREGADO` (referido como `e.ssn`) é igual ao `essn` da tabela `DEPENDENTE` (referido como `d.essn`). Essa condição é crucial para determinar como as tabelas são unidas. `ssn` e `essn` são assumidos como chaves que ligam empregados aos seus dependentes.

### **ORDER BY Clause**
```sql
ORDER BY e.ssn ASC, d.parentesco DESC;
```
- **ORDER BY** é usado para ordenar os resultados da consulta com base em uma ou mais colunas. Aqui, os resultados são primeiro ordenados pelo `ssn` do empregado em ordem ascendente (`ASC`), e então pelo `parentesco` em ordem descendente (`DESC`).
  - `ASC`: Ascendente, do menor para o maior ou de A a Z.
  - `DESC`: Descendente, do maior para o menor ou de Z a A.

b) Mostre o nome e endereço de todos os empregados que trabalham para o departamento
Pesquisa.

In [20]:
cursor.execute("SELECT e.pnome, e.endereco FROM EMPREGADO e JOIN DEPARTAMENTO d ON e.dno = d.dnumero WHERE d.dnome = 'Pesquisa';")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)


('Joao', 'R. 24 de maio, 1500 - Curitiba - PR')
('Fabio', 'R. Alagoas, 325 - Curitiba - PR')
('Ricardo', 'Av. Figueira, 55 - Almirante Tamandare - PR')
('Jussara', 'Trav. da Lapa, 23 - Curitiba - PR')


### **SELECT Clause**
```sql
SELECT e.pnome, e.endereco
```
- **SELECT**: Esta cláusula é usada para especificar quais colunas serão retornadas pelo resultado da consulta. 
  - `e.pnome`: O `pnome` da tabela `EMPREGADO`, onde `e` é o alias dado à tabela `EMPREGADO`. `pnome` representa o primeiro nome do empregado.
  - `e.endereco`: O `endereco` da tabela `EMPREGADO`. Este campo armazena o endereço do empregado.

### **FROM Clause**
```sql
FROM EMPREGADO e
```
- **FROM**: Indica de qual tabela os dados serão selecionados. A tabela `EMPREGADO` é aqui referida pelo alias `e`. Usar um alias simplifica referências subsequentes à tabela, especialmente útil em consultas com múltiplas tabelas ou junções.

### **JOIN Clause**
```sql
JOIN DEPARTAMENTO d ON e.dno = d.dnumero
```
- **JOIN**: Esta cláusula é utilizada para combinar registros de duas tabelas baseadas em uma condição relacionada. Neste caso, a tabela `DEPARTAMENTO` é trazida para a consulta e é dada o alias `d`.
  - **ON e.dno = d.dnumero**: Esta é a condição de junção que especifica como as tabelas `EMPREGADO` e `DEPARTAMENTO` são relacionadas. O `dno` (número do departamento) na tabela `EMPREGADO` deve corresponder ao `dnumero` (número do departamento) na tabela `DEPARTAMENTO`. Essencialmente, isso liga cada empregado ao departamento ao qual pertence.

### **WHERE Clause**
```sql
WHERE d.dnome = 'Pesquisa';
```
- **WHERE**: Esta cláusula é usada para filtrar registros e limitar quais linhas serão incluídas no resultado final da consulta. Neste caso, apenas os registros em que o `dnome` (nome do departamento) da tabela `DEPARTAMENTO` é igual a 'Pesquisa' serão retornados. Isso restringe a saída para incluir apenas empregados que trabalham no departamento chamado "Pesquisa".

c) Para todo projeto localizado em Araucaria, liste o nome do projeto, o nome do departamento
de controle e o último nome, endereço e data de nascimento do gerente do departamento.

In [24]:
cursor.execute("""SELECT p.pjnome AS "Nome do Projeto", d.dnome AS "Nome do Departamento", e.unome AS "Sobrenome", e.endereco AS "Endereço", e.datanasc AS "Data de Nascimento"
FROM PROJETO p
JOIN DEPARTAMENTO d ON p.dnum = d.dnumero
JOIN EMPREGADO e ON d.gerssn = e.ssn
WHERE p.plocal = 'Araucaria';""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('ProdutoY', 'Pesquisa', 'Souza', 'R. 24 de maio, 1500 - Curitiba - PR', datetime.date(1965, 1, 9))


**SELECT Clause**
   - `p.pjnome AS "Nome do Projeto"`: Seleciona o nome do projeto da tabela `PROJETO` e renomeia a coluna no resultado como "Nome do Projeto".
   - `d.dnome AS "Nome do Departamento"`: Seleciona o nome do departamento da tabela `DEPARTAMENTO` e renomeia a coluna no resultado como "Nome do Departamento".
   - `e.unome AS "Sobrenome"`: Seleciona o sobrenome do empregado da tabela `EMPREGADO` e renomeia a coluna no resultado como "Sobrenome".
   - `e.endereco AS "Endereço"`: Seleciona o endereço do empregado da tabela `EMPREGADO` e renomeia a coluna no resultado como "Endereço".
   - `e.datanasc AS "Data de Nascimento"`: Seleciona a data de nascimento do empregado da tabela `EMPREGADO` e renomeia a coluna no resultado como "Data de Nascimento".

**FROM Clause**
   - `FROM PROJETO p`: Especifica que os dados iniciais são retirados da tabela `PROJETO`, que é referenciada com o alias `p` para simplificar as referências subsequentes na consulta.

**JOIN Clauses**
   - `JOIN DEPARTAMENTO d ON p.dnum = d.dnumero`: Realiza um `INNER JOIN` com a tabela `DEPARTAMENTO`, que é referenciada com o alias `d`. A junção é feita com base na condição de que o número do departamento no projeto (`p.dnum`) corresponda ao número do departamento (`d.dnumero`), conectando assim cada projeto ao seu respectivo departamento.
   - `JOIN EMPREGADO e ON d.gerssn = e.ssn`: Realiza outro `INNER JOIN`, desta vez com a tabela `EMPREGADO`, referenciada como `e`. Esta junção é baseada na condição de que o SSN do gerente do departamento (`d.gerssn`) corresponda ao SSN do empregado (`e.ssn`), conectando o departamento ao seu gerente.

**WHERE Clause**
   - `WHERE p.plocal = 'Araucaria'`: Filtra os resultados para incluir apenas aqueles projetos que estão localizados em 'Araucaria'. Esta condição é aplicada após as junções serem feitas, restringindo ainda mais os dados que serão exibidos no resultado final.

d) Recupere os nomes de todos os empregados que trabalhem mais de 10 horas por semana
no projeto Automatizacao.

In [28]:
cursor.execute("""SELECT e.pnome AS "Nome do Empregado"
FROM EMPREGADO e
JOIN TRABALHA t ON e.ssn = t.essn
JOIN PROJETO p ON t.pno = p.pnumero
WHERE p.pjnome = 'Automatizacao' AND t.horas > 10;
""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Alberto',)


**SELECT Clause**
   - `e.pnome AS "Nome do Empregado"`: Esta parte da consulta especifica que o primeiro nome dos empregados (`pnome`) será selecionado da tabela `EMPREGADO`. O resultado exibirá esta coluna com o cabeçalho "Nome do Empregado". O uso do alias `e` para a tabela `EMPREGADO` facilita a referência à coluna dentro da consulta.

**FROM Clause**
   - `FROM EMPREGADO e`: Define que a tabela `EMPREGADO` é a tabela principal de onde os dados serão inicialmente retirados. O alias `e` é dado à tabela `EMPREGADO`, o que ajuda a simplificar as referências a essa tabela nas cláusulas subsequentes.

**JOIN Clauses**
   - `JOIN TRABALHA t ON e.ssn = t.essn`: Esta cláusula realiza uma junção entre a tabela `EMPREGADO` e a tabela `TRABALHA` usando o alias `t` para `TRABALHA`. A junção é feita baseada na correspondência entre `ssn` (Número do Seguro Social do empregado) em `EMPREGADO` e `essn` (Número do Seguro Social do empregado registrado em `TRABALHA`). Isso vincula cada registro de empregado ao seu registro correspondente na tabela `TRABALHA`.
   - `JOIN PROJETO p ON t.pno = p.pnumero`: Realiza outro `INNER JOIN`, desta vez com a tabela `PROJETO`, usando o alias `p`. Esta junção baseia-se na correspondência entre `pno` (número do projeto em `TRABALHA`) e `pnumero` (número do projeto em `PROJETO`), conectando as entradas de trabalho aos projetos específicos.

**WHERE Clause**
   - `WHERE p.pjnome = 'Automatizacao' AND t.horas > 10`: Esta cláusula filtra os registros para incluir apenas aqueles em que o nome do projeto é "Automatização" (`pjnome` em `PROJETO`) e o número de horas trabalhadas por semana é maior que 10 (`horas` em `TRABALHA`). Isso assegura que apenas empregados que atendem a ambos os critérios sejam incluídos nos resultados.

e) Mostre o nome dos empregados que têm os três maiores salários

In [33]:
cursor.execute("""SELECT pnome AS "Nome do Empregado"
FROM EMPREGADO
ORDER BY salario DESC
LIMIT 3;

""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Roberto',)
('Joaquim',)
('Jennifer',)


**SELECT Clause**
   - `pnome AS "Nome do Empregado"`: Esta parte da consulta especifica que o primeiro nome dos empregados (`pnome`) será selecionado da tabela `EMPREGADO`. O resultado exibirá esta coluna com o cabeçalho "Nome do Empregado". Isso facilita a leitura e interpretação dos dados no resultado, proporcionando clareza sobre qual informação está sendo apresentada.

**FROM Clause**
   - `FROM EMPREGADO`: Esta cláusula especifica a tabela de onde os dados serão extraídos. Neste caso, a tabela é `EMPREGADO`. Não há alias usado aqui, pois a consulta é direta e envolve apenas uma tabela.

**ORDER BY Clause**
   - `ORDER BY salario DESC`: Esta cláusula é usada para ordenar os dados retornados pela consulta com base na coluna `salario`. A palavra-chave `DESC` indica que a ordenação deve ser descendente, ou seja, do maior para o menor. Isso garante que os empregados com os maiores salários apareçam primeiro nos resultados da consulta.

**LIMIT Clause**
   - `LIMIT 3`: Esta cláusula limita o número de registros retornados pela consulta a apenas três. Isso é útil quando você está interessado apenas em uma pequena quantidade de registros de topo baseados em algum critério, como os maiores salários neste caso. O uso do `LIMIT` é uma maneira eficiente de restringir a quantidade de dados processados e apresentados, especialmente em tabelas grandes.

f) Mostre o nome e o salário dos supervisores com aumento de 20%.

In [34]:
cursor.execute("""SELECT pnome AS "Nome do Empregado", salario * 1.20 AS "Salario Aumentado"
FROM EMPREGADO
WHERE ssn IN (SELECT superssn FROM EMPREGADO WHERE superssn IS NOT NULL);
""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Fabio', 48000.0)
('Jennifer', 51600.0)
('Joaquim', 66000.0)
('Roberto', 69600.0)


**SELECT Clause**
   - `pnome AS "Nome do Empregado"`: Esta parte da consulta seleciona o campo `pnome` da tabela `EMPREGADO`, que contém o primeiro nome dos empregados. O nome da coluna no resultado é renomeado para "Nome do Empregado" usando a cláusula `AS`. Isso torna os resultados mais compreensíveis, especialmente em contextos onde o relatório final é apresentado a stakeholders.
   - `salario * 1.20 AS "Salario Aumentado"`: Aqui, o salário registrado na tabela para cada empregado é aumentado em 20% através de uma operação aritmética simples (`salario * 1.20`). O resultado dessa operação é apresentado sob o rótulo "Salario Aumentado". Esta coluna no resultado mostra o novo salário que um empregado receberia se um aumento de 20% fosse aplicado, útil para análises de ajustes salariais ou simulações de aumento de custos.

**FROM Clause**
   - `FROM EMPREGADO`: Esta cláusula especifica que os dados são retirados da tabela `EMPREGADO`. Esta é a tabela principal onde os dados sobre os empregados, incluindo seus salários e números de seguro social (SSN), são armazenados.

**WHERE Clause**
   - `WHERE ssn IN (SELECT superssn FROM EMPREGADO WHERE superssn IS NOT NULL)`: Esta parte da consulta é crucial para filtrar apenas os empregados que são supervisores. 
     - `SELECT superssn FROM EMPREGADO WHERE superssn IS NOT NULL`: Esta subconsulta seleciona todos os `superssn` (número de seguro social do supervisor) da tabela `EMPREGADO` que não são nulos. O `superssn` é o campo que identifica um empregado como supervisor de outros empregados.
     - `WHERE ssn IN (...)`: A cláusula principal `WHERE` usa o resultado da subconsulta para filtrar empregados cujos `ssn` estão na lista de `superssn` obtida. Isso garante que apenas os registros de empregados que são supervisores sejam incluídos nos resultados finais.

g) Mostre os departamentos e suas localidades.

In [38]:
cursor.execute("""SELECT d.dnome AS "Nome do Departamento", l.dlocalizacao AS "Localização do Departamento"
FROM DEPARTAMENTO d
JOIN LOCALIZACAO l ON d.dnumero = l.dnum;
""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Sede Administrativa', 'Curitiba')
('Administracao', 'Colombo')
('Pesquisa', 'Araucaria')
('Pesquisa', 'Curitiba')
('Pesquisa', 'Pinhais')


**SELECT Clause**
   - `d.dnome AS "Nome do Departamento"`: Esta parte da consulta seleciona o campo `dnome` da tabela `DEPARTAMENTO`, que contém os nomes dos departamentos. O resultado dessa seleção é apresentado no conjunto de resultados com o cabeçalho "Nome do Departamento". A utilização do alias `d` para a tabela `DEPARTAMENTO` facilita a leitura e referência ao campo dentro da consulta.
   - `l.dlocalizacao AS "Localização do Departamento"`: Similarmente, esta parte seleciona a coluna `dlocalizacao` da tabela `LOCALIZACAO`, que contém as localizações geográficas ou físicas dos departamentos. A coluna é renomeada para "Localização do Departamento" no resultado, proporcionando clareza e tornando os dados mais acessíveis para quem visualiza o resultado.

**FROM Clause**
   - `FROM DEPARTAMENTO d`: Esta cláusula especifica a tabela de onde os dados principais são retirados, no caso, a tabela `DEPARTAMENTO`. O alias `d` é atribuído a esta tabela, simplificando as referências subsequentes na consulta.

**JOIN Clause**
   - `JOIN LOCALIZACAO l ON d.dnumero = l.dnum`: Aqui, a tabela `LOCALIZACAO` é juntada à consulta usando um `INNER JOIN`. O `INNER JOIN` é realizado com base na condição especificada pelo `ON d.dnumero = l.dnum`. Isso significa que a junção é feita onde o número do departamento na tabela `DEPARTAMENTO` (`d.dnumero`) corresponde ao número do departamento na tabela `LOCALIZACAO` (`l.dnum`). Isso garante que cada nome de departamento será associado corretamente à sua localização específica.

h) Mostre os departamentos e seus projetos.

In [39]:
cursor.execute("""SELECT d.dnome AS "Nome do Departamento", p.pjnome AS "Nome do Projeto"
FROM DEPARTAMENTO d
JOIN PROJETO p ON d.dnumero = p.dnum;
""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Pesquisa', 'ProdutoX')
('Pesquisa', 'ProdutoY')
('Pesquisa', 'ProdutoZ')
('Administracao', 'Automatizacao')
('Sede Administrativa', 'Reorganizacao')
('Administracao', 'Novos Beneficios')


**SELECT Clause**
   - `d.dnome AS "Nome do Departamento"`: Esta parte da consulta seleciona a coluna `dnome` da tabela `DEPARTAMENTO`, que contém os nomes dos departamentos. A coluna é renomeada como "Nome do Departamento" no resultado final usando a cláusula `AS`. Isso ajuda a tornar os dados mais compreensíveis ao serem visualizados, especialmente em relatórios ou interfaces de usuário onde a clareza é crucial.
   - `p.pjnome AS "Nome do Projeto"`: Similarmente, esta parte seleciona a coluna `pjnome` da tabela `PROJETO`, que contém os nomes dos projetos. A coluna é renomeada como "Nome do Projeto" no resultado final. Isso fornece um mapeamento direto e claro do nome de cada projeto ao departamento correspondente.

**FROM Clause**
   - `FROM DEPARTAMENTO d`: Especifica a tabela `DEPARTAMENTO` como a fonte dos dados, com `d` servindo como um alias para `DEPARTAMENTO`. Usar aliases é uma prática comum para simplificar a escrita e interpretação das consultas, especialmente quando múltiplas tabelas estão envolvidas.

**JOIN Clause**
   - `JOIN PROJETO p ON d.dnumero = p.dnum`: Esta cláusula executa um `INNER JOIN` entre as tabelas `DEPARTAMENTO` e `PROJETO`. O `INNER JOIN` é feito com base na condição que o número do departamento (`d.dnumero` na tabela `DEPARTAMENTO`) coincide com o número do departamento referenciado na tabela `PROJETO` (`p.dnum`). Este método de junção garante que apenas os pares de departamentos e projetos que têm uma correspondência direta entre esses campos específicos serão incluídos nos resultados finais.

i) Mostre os empregados do sexo feminino e a data de nascimento cujo salário é maior que
30.000.

In [40]:
cursor.execute("""SELECT pnome AS "Nome do Empregado", datanasc AS "Data de Nascimento"
FROM EMPREGADO
WHERE sexo = 'F' AND salario > 30000;
""")
# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Jennifer', datetime.date(1941, 6, 20))


**SELECT Clause**
   - `pnome AS "Nome do Empregado"`: Esta parte da consulta seleciona a coluna `pnome` da tabela `EMPREGADO`. `pnome` representa o primeiro nome dos empregados. A coluna é renomeada para "Nome do Empregado" no resultado final usando a cláusula `AS`, o que ajuda a tornar os resultados mais claros e fáceis de entender para quem está visualizando os dados.
   - `datanasc AS "Data de Nascimento"`: Da mesma forma, esta cláusula seleciona a coluna `datanasc`, que contém as datas de nascimento dos empregados. Ela é renomeada para "Data de Nascimento" no resultado final, proporcionando uma visualização clara da informação.

**FROM Clause**
   - `FROM EMPREGADO`: Esta cláusula especifica que os dados serão retirados da tabela `EMPREGADO`. Esta tabela contém informações detalhadas sobre os empregados de uma organização, incluindo nomes, datas de nascimento, gênero, salários, entre outros dados pessoais e profissionais.

**WHERE Clause**
   - `WHERE sexo = 'F' AND salario > 30000`: Esta parte da consulta é usada para filtrar os registros com base em duas condições:
     - `sexo = 'F'`: Filtra os empregados pelo gênero, incluindo apenas aqueles do sexo feminino (`F`).
     - `salario > 30000`: Adicionalmente, a consulta filtra para incluir somente os empregados cujos salários são superiores a 30.000. Este é um critério numérico que seleciona empregados com remuneração acima de um certo limiar.

j) Mostre os projetos em que o empregado 'Fábio' trabalha.

In [46]:
cursor.execute("""SELECT p.pjnome AS "Nome do Projeto"
FROM PROJETO p
JOIN TRABALHA t ON p.pnumero = t.pno
JOIN EMPREGADO e ON t.essn = e.ssn
WHERE e.pnome = 'Fabio';
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

### Análise dos Dados
- Da tabela `EMPREGADO`, confirmamos que 'Fabio' está listado com um `SSN` de `'333445555'` e trabalha no departamento número `'5'`, com um salário de `40000.0`.
- Da tabela `TRABALHA`, não vemos nenhum registro diretamente listado para 'Fabio' (`333445555`), o que sugere por que a consulta original não retornou resultados.

### Verificação do Relacionamento entre Fabio e Projetos
- Observamos que, apesar de Fabio estar listado na tabela `EMPREGADO`, não há entradas correspondentes para ele na tabela `TRABALHA` baseadas nos dados mostrados. Isso indica que, ou 'Fabio' não está atualmente alocado a nenhum projeto, já que removemos as ocorrências na atividade 1 f).

**SELECT Clause**
   - `p.pjnome AS "Nome do Projeto"`: Esta cláusula seleciona o nome do projeto (`pjnome`) da tabela `PROJETO`, onde `p` é o alias utilizado para a tabela `PROJETO`. O resultado dessa seleção é renomeado para "Nome do Projeto" para apresentação nos resultados finais. Isso torna claro o que cada valor na coluna representa, facilitando a interpretação dos dados.

**FROM Clause**
   - `FROM PROJETO p`: Esta cláusula especifica que os dados principais estão sendo extraídos da tabela `PROJETO`, com `p` servindo como um alias para simplificar referências na consulta.

**JOIN Clauses**
   - `JOIN TRABALHA t ON p.pnumero = t.pno`: Esta primeira junção conecta a tabela `PROJETO` com a tabela `TRABALHA`. O `ON p.pnumero = t.pno` especifica que a junção deve ser feita onde o número do projeto na tabela `PROJETO` (`p.pnumero`) corresponde ao número do projeto na tabela `TRABALHA` (`t.pno`). Esta junção permite relacionar projetos específicos com as entradas correspondentes na tabela `TRABALHA`, que registra os detalhes sobre o trabalho dos empregados nesses projetos.
   - `JOIN EMPREGADO e ON t.essn = e.ssn`: A segunda junção conecta a tabela `TRABALHA` à tabela `EMPREGADO`. Aqui, a condição `ON t.essn = e.ssn` especifica que a junção deve ocorrer onde o número de seguro social do empregado registrado em `TRABALHA` (`t.essn`) corresponde ao número de seguro social (`ssn`) na tabela `EMPREGADO`. Isso liga as informações de trabalho à tabela de empregados, associando cada entrada de trabalho ao empregado correspondente.

**WHERE Clause**
   - `WHERE e.pnome = 'Fabio';`: Esta cláusula filtra os resultados para incluir apenas aqueles registros onde o primeiro nome do empregado é 'Fabio'. O filtro é aplicado após as junções terem sido realizadas, restringindo a saída final aos projetos em que 'Fabio' está envolvido. Esta cláusula é essencial para garantir que apenas os dados relevantes para o empregado especificado sejam retornados.


# Atividade 3) Funções agregadas e nativas - 

Utilize funções agregadas e/ou nativas
SQL para responder as seguintes questões:


a) Recupere a média salarial de todos os empregados do sexo feminino.

In [6]:
cursor.execute("""SELECT AVG(salario) AS media_salarial_feminina
FROM EMPREGADO
WHERE sexo = 'F';""")


# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

(31000.0,)


**SELECT Clause**
   - `SELECT AVG(salario) AS media_salarial_feminina`: Nesta parte da consulta, estamos utilizando a função agregada `AVG()` para calcular a média dos salários de uma determinada seleção de empregados. A coluna `salario` da tabela `EMPREGADO` é utilizada para este cálculo. O resultado da média é renomeado para `media_salarial_feminina` usando a cláusula `AS`, o que facilita a interpretação do resultado, indicando claramente que se trata da média salarial de empregadas do sexo feminino.

**FROM Clause**
   - `FROM EMPREGADO`: Esta cláusula indica que os dados utilizados para o cálculo da média salarial serão extraídos da tabela `EMPREGADO`. A tabela `EMPREGADO` contém informações sobre os empregados, como nome, data de nascimento, gênero, salário, entre outros detalhes relevantes.

**WHERE Clause**
   - `WHERE sexo = 'F'`: A cláusula `WHERE` é usada para filtrar os registros da tabela `EMPREGADO`, selecionando apenas aqueles onde a coluna `sexo` é igual a 'F'. Isso significa que apenas as empregadas do sexo feminino serão consideradas no cálculo da média salarial. Este filtro é crucial para garantir que a média seja calculada exclusivamente para o grupo de interesse (empregadas do sexo feminino).

b) Mostre o número de empregados por supervisor.

In [7]:
cursor.execute("""SELECT superssn AS supervisor_ssn, COUNT(*) AS numero_de_empregados
FROM EMPREGADO
GROUP BY superssn;""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

(None, 1)
('888665555', 3)
('987654321', 2)
('333445555', 3)


**SELECT Clause**
   - `superssn AS supervisor_ssn`: Esta parte da consulta seleciona a coluna `superssn` da tabela `EMPREGADO`, que representa o número de segurança social do supervisor de cada empregado. Renomeamos essa coluna para `supervisor_ssn` no resultado final para facilitar a compreensão de que este valor se refere ao supervisor.
   - `COUNT(*) AS numero_de_empregados`: A função `COUNT(*)` conta o número total de registros (empregados) que estão sendo agrupados. Renomeamos o resultado dessa contagem para `numero_de_empregados`, indicando claramente que o valor representa a quantidade de empregados sob a supervisão de cada supervisor.

**FROM Clause**
   - `FROM EMPREGADO`: Esta cláusula especifica que os dados serão extraídos da tabela `EMPREGADO`, onde estão armazenadas as informações sobre os empregados e seus supervisores.

**GROUP BY Clause**
   - `GROUP BY superssn`: A cláusula `GROUP BY` agrupa os registros da tabela `EMPREGADO` com base na coluna `superssn`. Isso significa que a contagem dos empregados será feita para cada grupo de empregados que compartilham o mesmo supervisor (identificado pelo número de segurança social do supervisor).

c) Mostre o maior número de horas envolvido em projetos.

In [9]:
cursor.execute("""SELECT MAX(horas) AS max_horas
FROM TRABALHA;""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

(40.0,)


**SELECT Clause**
   - `MAX(horas) AS max_horas`: A função `MAX()` é usada para encontrar o maior valor na coluna `horas`, que representa o número de horas que um empregado trabalhou em um projeto. O resultado é renomeado como `max_horas` para facilitar a interpretação do resultado, indicando que este valor é o máximo número de horas trabalhadas em qualquer projeto.

**FROM Clause**
   - `FROM TRABALHA`: Esta cláusula especifica que os dados utilizados para encontrar o valor máximo serão retirados da tabela `TRABALHA`. A tabela `TRABALHA` contém informações sobre as horas que cada empregado dedica a diferentes projetos.

d) Para cada projeto, liste o nome do projeto e o total de horas por semana (de todos os
empregados) gastas no projeto.

In [10]:
cursor.execute("""SELECT PROJETO.pjnome AS nome_do_projeto, SUM(TRABALHA.horas) AS total_horas_por_semana
FROM PROJETO
JOIN TRABALHA ON PROJETO.pnumero = TRABALHA.pno
GROUP BY PROJETO.pjnome;""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Automatizacao', 45.0)
('ProdutoZ', 40.0)
('ProdutoY', 27.5)
('Novos Beneficios', 55.0)
('ProdutoX', 52.5)
('Reorganizacao', 15.0)


**SELECT Clause**
   - `PROJETO.pjnome AS nome_do_projeto`: Esta parte da consulta seleciona a coluna `pjnome` da tabela `PROJETO`, que contém o nome do projeto. Renomeamos essa coluna como `nome_do_projeto` para deixar claro no resultado final que se trata do nome de cada projeto.
   - `SUM(TRABALHA.horas) AS total_horas_por_semana`: A função `SUM()` é usada para somar todas as horas que os empregados dedicaram a cada projeto. O resultado dessa soma é renomeado como `total_horas_por_semana`, indicando o total de horas gastas por semana em cada projeto.

**FROM Clause**
   - `FROM PROJETO`: Especifica que a consulta vai começar a selecionar dados da tabela `PROJETO`.

**JOIN Clause**
   - `JOIN TRABALHA ON PROJETO.pnumero = TRABALHA.pno`: A cláusula `JOIN` une as tabelas `PROJETO` e `TRABALHA` com base na relação entre `pnumero` (da tabela `PROJETO`) e `pno` (da tabela `TRABALHA`). Isso permite combinar informações sobre projetos com os dados de horas trabalhadas por cada empregado nesses projetos.

**GROUP BY Clause**
   - `GROUP BY PROJETO.pjnome`: A cláusula `GROUP BY` agrupa os resultados pelo nome do projeto (`pjnome`), garantindo que a soma das horas (`SUM(horas)`) seja calculada separadamente para cada projeto.

e) Para cada departamento, recupere o nome do departamento e a média salarial de todos os
empregados que trabalham nesse departamento.

In [11]:
cursor.execute("""SELECT DEPARTAMENTO.dnome AS nome_do_departamento, AVG(EMPREGADO.salario) AS media_salarial
FROM DEPARTAMENTO
JOIN EMPREGADO ON DEPARTAMENTO.dnumero = EMPREGADO.dno
GROUP BY DEPARTAMENTO.dnome;""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Administracao', 31000.0)
('Pesquisa', 33250.0)
('Sede Administrativa', 56500.0)


**SELECT Clause**
   - `DEPARTAMENTO.dnome AS nome_do_departamento`: Esta parte da consulta seleciona a coluna `dnome` da tabela `DEPARTAMENTO`, que contém o nome de cada departamento. Renomeamos essa coluna como `nome_do_departamento` para tornar claro no resultado final que se trata do nome do departamento.
   - `AVG(EMPREGADO.salario) AS media_salarial`: A função `AVG()` é usada para calcular a média dos salários de todos os empregados que trabalham em cada departamento. O resultado é renomeado como `media_salarial`, indicando que se trata da média salarial dos empregados em cada departamento.

**FROM Clause**
   - `FROM DEPARTAMENTO`: Especifica que a consulta começa a selecionar dados da tabela `DEPARTAMENTO`.

**JOIN Clause**
   - `JOIN EMPREGADO ON DEPARTAMENTO.dnumero = EMPREGADO.dno`: A cláusula `JOIN` une as tabelas `DEPARTAMENTO` e `EMPREGADO` com base na relação entre `dnumero` (da tabela `DEPARTAMENTO`) e `dno` (da tabela `EMPREGADO`). Isso permite combinar informações sobre departamentos com os dados dos empregados que trabalham neles.

**GROUP BY Clause**
   - `GROUP BY DEPARTAMENTO.dnome`: A cláusula `GROUP BY` agrupa os resultados pelo nome do departamento (`dnome`), garantindo que a média salarial (`AVG(salario)`) seja calculada separadamente para cada departamento.

f) Liste os nomes de todos os empregados com dois ou mais dependentes

In [12]:
cursor.execute("""SELECT EMPREGADO.pnome AS primeiro_nome, EMPREGADO.unome AS ultimo_nome
FROM EMPREGADO
JOIN DEPENDENTE ON EMPREGADO.ssn = DEPENDENTE.essn
GROUP BY EMPREGADO.ssn, EMPREGADO.pnome, EMPREGADO.unome
HAVING COUNT(DEPENDENTE.nomedep) >= 2;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Joao', 'Souza')
('Fabio', 'Will')


**SELECT Clause**
   - `EMPREGADO.pnome AS primeiro_nome, EMPREGADO.unome AS ultimo_nome`: Esta parte da consulta seleciona os nomes dos empregados (`pnome` e `unome`) das colunas da tabela `EMPREGADO`. Esses campos são renomeados como `primeiro_nome` e `ultimo_nome`, respectivamente, para tornar os resultados mais claros.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta começará a selecionar dados da tabela `EMPREGADO`.

**JOIN Clause**
   - `JOIN DEPENDENTE ON EMPREGADO.ssn = DEPENDENTE.essn`: A cláusula `JOIN` une as tabelas `EMPREGADO` e `DEPENDENTE` com base na relação entre o número de segurança social do empregado (`ssn`) na tabela `EMPREGADO` e o número de segurança social do empregado referenciado (`essn`) na tabela `DEPENDENTE`. Isso permite combinar informações sobre empregados com seus dependentes.

**GROUP BY Clause**
   - `GROUP BY EMPREGADO.ssn, EMPREGADO.pnome, EMPREGADO.unome`: A cláusula `GROUP BY` agrupa os registros por empregado (identificado pelo número de segurança social `ssn`) e seus nomes (`pnome`, `unome`). Isso é necessário para garantir que a contagem de dependentes seja realizada por empregado.

**HAVING Clause**
   - `HAVING COUNT(DEPENDENTE.nomedep) >= 2`: A cláusula `HAVING` filtra os grupos (empregados) que têm dois ou mais dependentes, contados pela coluna `nomedep` na tabela `DEPENDENTE`. Apenas os empregados que têm dois ou mais dependentes serão incluídos nos resultados.

g) Mostre o nome do departamento que possui o menor número de projetos associados.

In [13]:
cursor.execute("""SELECT DEPARTAMENTO.dnome AS nome_do_departamento, COUNT(PROJETO.pnumero) AS numero_de_projetos
FROM DEPARTAMENTO
JOIN PROJETO ON DEPARTAMENTO.dnumero = PROJETO.dnum
GROUP BY DEPARTAMENTO.dnome
ORDER BY numero_de_projetos ASC
LIMIT 1;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Sede Administrativa', 1)


**SELECT Clause**
   - `DEPARTAMENTO.dnome AS nome_do_departamento`: Esta parte da consulta seleciona o nome do departamento (`dnome`) da tabela `DEPARTAMENTO` e o renomeia como `nome_do_departamento` para facilitar a interpretação no resultado final.
   - `COUNT(PROJETO.pnumero) AS numero_de_projetos`: A função `COUNT()` é usada para contar o número de projetos (`pnumero`) associados a cada departamento. O resultado é renomeado como `numero_de_projetos`, indicando o número total de projetos por departamento.

**FROM Clause**
   - `FROM DEPARTAMENTO`: Especifica que a consulta começa a selecionar dados da tabela `DEPARTAMENTO`.

**JOIN Clause**
   - `JOIN PROJETO ON DEPARTAMENTO.dnumero = PROJETO.dnum`: A cláusula `JOIN` une as tabelas `DEPARTAMENTO` e `PROJETO` com base na relação entre `dnumero` (da tabela `DEPARTAMENTO`) e `dnum` (da tabela `PROJETO`). Isso permite combinar informações sobre departamentos com os projetos a eles associados.

**GROUP BY Clause**
   - `GROUP BY DEPARTAMENTO.dnome`: A cláusula `GROUP BY` agrupa os resultados pelo nome do departamento (`dnome`), garantindo que a contagem dos projetos (`COUNT(pnumero)`) seja calculada separadamente para cada departamento.

**ORDER BY Clause**
   - `ORDER BY numero_de_projetos ASC`: A cláusula `ORDER BY` ordena os resultados pela contagem de projetos (`numero_de_projetos`) em ordem crescente (`ASC`), colocando o departamento com o menor número de projetos no topo da lista.

**LIMIT Clause**
   - `LIMIT 1`: A cláusula `LIMIT 1` garante que apenas o departamento com o menor número de projetos seja retornado.

h) Escreva uma consulta que retorne do 10o ao 22o caractere do endereço do empregado.

In [16]:
cursor.execute("""SELECT SUBSTRING(endereco, 10, 13) AS trecho_endereco
FROM EMPREGADO;""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('maio, 1500 - ',)
('s, 325 - Curi',)
('a, 98 - Pinha',)
(' 291 - Colomb',)
('ira, 55 - Alm',)
('Lapa, 23 - Cu',)
(' 1245 - Curit',)
(' 450 - Curiti',)
(' – Araucária ',)


**SELECT Clause**
   - `SUBSTRING(endereco, 10, 13) AS trecho_endereco`: A função `SUBSTRING()` é usada para extrair uma parte específica de uma string. 
     - O primeiro argumento, `endereco`, especifica a coluna de onde queremos extrair a string.
     - O segundo argumento, `10`, indica a posição inicial do caractere (10º caractere).
     - O terceiro argumento, `13`, especifica o número de caracteres a serem extraídos, ou seja, do 10º ao 22º caractere. O valor 13 é usado porque o intervalo de caracteres de 10 a 22 inclui 13 caracteres no total.
     - O resultado é renomeado como `trecho_endereco` para facilitar a compreensão de que essa parte corresponde a um trecho do endereço.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta está selecionando dados da tabela `EMPREGADO`.

i) Escreva uma consulta que retorne apenas o mês de nascimento de cada funcionário.

In [25]:
cursor.execute("""SELECT pnome AS primeiro_nome, unome AS ultimo_nome, EXTRACT(MONTH FROM datanasc) AS mes_nascimento
FROM EMPREGADO;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Joao', 'Souza', Decimal('1'))
('Fabio', 'Will', Decimal('12'))
('Alice', 'Zebra', Decimal('1'))
('Jennifer', 'Wallace', Decimal('6'))
('Ricardo', 'Nantes', Decimal('9'))
('Jussara', 'Pereira', Decimal('7'))
('Alberto', 'Medeiros', Decimal('3'))
('Joaquim', 'Brito', Decimal('11'))
('Roberto', 'Silva', Decimal('6'))


**SELECT Clause**
   - `pnome AS primeiro_nome`: Esta parte da consulta seleciona a coluna `pnome` da tabela `EMPREGADO`, que contém o primeiro nome dos funcionários. A coluna é renomeada como `primeiro_nome` para tornar os resultados mais claros e fáceis de entender.
   - `unome AS ultimo_nome`: Seleciona a coluna `unome`, que contém o sobrenome dos funcionários. Ela é renomeada como `ultimo_nome`, proporcionando uma visualização clara do nome completo do funcionário no resultado.
   - `EXTRACT(MONTH FROM datanasc) AS mes_nascimento`: Utiliza a função `EXTRACT()` para extrair o mês da data de nascimento (`datanasc`) de cada funcionário. A parte `MONTH` especifica que estamos interessados apenas no mês da data, e o resultado é renomeado como `mes_nascimento` para indicar claramente que se trata do mês de nascimento.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta irá obter os dados da tabela `EMPREGADO`, que contém as informações dos funcionários, como seus nomes e datas de nascimento.

j) Escreva uma consulta que retorne a idade (em anos) que o empregado tinha quando o
dependente de parentesco filhou ou filha nasceu.

In [35]:
cursor.execute("""SELECT EMPREGADO.pnome AS primeiro_nome, EMPREGADO.unome AS ultimo_nome, DEPENDENTE.nomedep AS nome_dependente,
       DEPENDENTE.parentesco, 
       EXTRACT(YEAR FROM AGE(DEPENDENTE.datanascdep, EMPREGADO.datanasc)) AS idade_quando_dependente_nasceu
FROM EMPREGADO
JOIN DEPENDENTE ON EMPREGADO.ssn = DEPENDENTE.essn
WHERE UPPER(DEPENDENTE.parentesco) IN ('FILHO', 'FILHA');
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Joao', 'Souza', 'Alice', 'FILHA', Decimal('23'))
('Joao', 'Souza', 'Michel', 'FILHO', Decimal('22'))
('Fabio', 'Will', 'Teodoro', 'FILHO', Decimal('27'))
('Fabio', 'Will', 'Alice', 'FILHA', Decimal('30'))


**SELECT Clause**
   - `EMPREGADO.pnome AS primeiro_nome`: Esta parte da consulta seleciona a coluna `pnome` da tabela `EMPREGADO`, que contém o primeiro nome dos empregados. A coluna é renomeada como `primeiro_nome` para facilitar a compreensão dos resultados.
   - `EMPREGADO.unome AS ultimo_nome`: Seleciona a coluna `unome`, que contém o sobrenome dos empregados. Ela é renomeada como `ultimo_nome` para que o resultado mostre claramente o nome completo dos empregados.
   - `DEPENDENTE.nomedep AS nome_dependente`: Seleciona a coluna `nomedep` da tabela `DEPENDENTE`, que contém o nome do dependente. A coluna é renomeada como `nome_dependente` para facilitar a interpretação.
   - `DEPENDENTE.parentesco`: Seleciona a coluna `parentesco`, que indica o relacionamento do dependente com o empregado (como "FILHO" ou "FILHA").
   - `EXTRACT(YEAR FROM AGE(DEPENDENTE.datanascdep, EMPREGADO.datanasc)) AS idade_quando_dependente_nasceu`: Utiliza a função `AGE()` para calcular a diferença entre a data de nascimento do dependente (`datanascdep`) e a data de nascimento do empregado (`datanasc`). A função `EXTRACT(YEAR FROM AGE(...))` extrai a diferença em anos, representando a idade do empregado quando o dependente nasceu. O resultado é renomeado como `idade_quando_dependente_nasceu`.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta seleciona dados da tabela `EMPREGADO`, que contém informações sobre os empregados.

**JOIN Clause**
   - `JOIN DEPENDENTE ON EMPREGADO.ssn = DEPENDENTE.essn`: Une as tabelas `EMPREGADO` e `DEPENDENTE` com base na correspondência entre o número de segurança social do empregado (`ssn`) e o número de segurança social referenciado na tabela `DEPENDENTE` (`essn`). Isso permite combinar informações dos empregados com seus dependentes.

**WHERE Clause**
   - `WHERE UPPER(DEPENDENTE.parentesco) IN ('FILHO', 'FILHA')`: Aplica um filtro para incluir apenas os dependentes cujo relacionamento com o empregado é "FILHO" ou "FILHA", independentemente de como o texto foi armazenado no banco de dados. A função `UPPER()` converte o valor da coluna `parentesco` para maiúsculas, garantindo uma comparação precisa.

k) Escreva uma consulte que conte o número de dependentes por ano de nascimento.

In [36]:
cursor.execute("""SELECT EXTRACT(YEAR FROM datanascdep) AS ano_de_nascimento, COUNT(*) AS numero_de_dependentes
FROM DEPENDENTE
GROUP BY EXTRACT(YEAR FROM datanascdep)
ORDER BY ano_de_nascimento;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

(Decimal('1942'), 1)
(Decimal('1958'), 1)
(Decimal('1967'), 1)
(Decimal('1970'), 1)
(Decimal('1983'), 1)
(Decimal('1986'), 1)
(Decimal('1988'), 2)


**SELECT Clause**
   - `EXTRACT(YEAR FROM datanascdep) AS ano_de_nascimento`: Utiliza a função `EXTRACT()` para extrair o ano da coluna `datanascdep`, que contém a data de nascimento dos dependentes. O resultado é renomeado como `ano_de_nascimento`, indicando que esta coluna contém o ano de nascimento dos dependentes.
   - `COUNT(*) AS numero_de_dependentes`: A função `COUNT(*)` conta o número total de registros (dependentes) para cada ano de nascimento. O resultado é renomeado como `numero_de_dependentes` para indicar que essa coluna contém a contagem de dependentes para cada ano.

**FROM Clause**
   - `FROM DEPENDENTE`: Especifica que a consulta irá obter os dados da tabela `DEPENDENTE`, que contém informações sobre os dependentes dos empregados.

**GROUP BY Clause**
   - `GROUP BY EXTRACT(YEAR FROM datanascdep)`: Agrupa os registros da tabela `DEPENDENTE` pelo ano de nascimento extraído da coluna `datanascdep`. Isso permite que a função `COUNT()` conte o número de dependentes para cada ano de nascimento.

**ORDER BY Clause**
   - `ORDER BY ano_de_nascimento`: Ordena os resultados em ordem crescente de ano de nascimento, para facilitar a visualização cronológica.

l) Escreva uma consulte que mostre o nome de empregados supervisor que tenham 2 ou mais
supervisionados.

In [37]:
cursor.execute("""SELECT EMPREGADO.pnome AS primeiro_nome, EMPREGADO.unome AS ultimo_nome, COUNT(SUBORDINADO.ssn) AS numero_de_supervisionados
FROM EMPREGADO
JOIN EMPREGADO AS SUBORDINADO ON EMPREGADO.ssn = SUBORDINADO.superssn
GROUP BY EMPREGADO.ssn, EMPREGADO.pnome, EMPREGADO.unome
HAVING COUNT(SUBORDINADO.ssn) >= 2;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Joaquim', 'Brito', 3)
('Jennifer', 'Wallace', 2)
('Fabio', 'Will', 3)


**SELECT Clause**
   - `EMPREGADO.pnome AS primeiro_nome, EMPREGADO.unome AS ultimo_nome`: Seleciona as colunas `pnome` e `unome` da tabela `EMPREGADO`, que contêm o primeiro nome e o último nome dos supervisores. Estas colunas são renomeadas como `primeiro_nome` e `ultimo_nome` para tornar os resultados mais claros.
   - `COUNT(SUBORDINADO.ssn) AS numero_de_supervisionados`: Utiliza a função `COUNT()` para contar o número de supervisionados (subordinados) de cada supervisor. A contagem é renomeada como `numero_de_supervisionados`.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta começa selecionando dados da tabela `EMPREGADO`.

**JOIN Clause**
   - `JOIN EMPREGADO AS SUBORDINADO ON EMPREGADO.ssn = SUBORDINADO.superssn`: Realiza um auto-join na tabela `EMPREGADO`, onde a tabela é unida a si mesma. A condição de junção é que o número de segurança social do supervisor (`ssn`) na tabela `EMPREGADO` corresponda ao número de segurança social do supervisor (`superssn`) na tabela `SUBORDINADO` (que é uma segunda instância da tabela `EMPREGADO`). Isso relaciona cada supervisor com seus supervisionados.

**GROUP BY Clause**
   - `GROUP BY EMPREGADO.ssn, EMPREGADO.pnome, EMPREGADO.unome`: Agrupa os registros por supervisor, identificado por `ssn`, e também agrupa por `pnome` e `unome` para garantir que cada supervisor seja tratado individualmente.

**HAVING Clause**
   - `HAVING COUNT(SUBORDINADO.ssn) >= 2`: Aplica um filtro para incluir apenas os supervisores que têm 2 ou mais supervisionados, conforme determinado pela contagem (`COUNT(SUBORDINADO.ssn)`).

m) Escreva uma consulta que mostre o valor mensal a ser pago por projeto (considere que a
coluna ‘salário’ de empregado é mensal).

In [38]:
cursor.execute("""SELECT PROJETO.pjnome AS nome_do_projeto, SUM(EMPREGADO.salario * TRABALHA.horas / 160) AS valor_mensal_projeto
FROM EMPREGADO
JOIN TRABALHA ON EMPREGADO.ssn = TRABALHA.essn
JOIN PROJETO ON TRABALHA.pno = PROJETO.pnumero
GROUP BY PROJETO.pjnome;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Automatizacao', 7031.25)
('ProdutoZ', 9500.0)
('ProdutoY', 4531.25)
('Novos Beneficios', 10843.75)
('ProdutoX', 9218.75)
('Reorganizacao', 4031.25)


**SELECT Clause**
   - `PROJETO.pjnome AS nome_do_projeto`: Seleciona a coluna `pjnome` da tabela `PROJETO`, que contém o nome do projeto. Renomeamos essa coluna como `nome_do_projeto` para tornar os resultados mais claros.
   - `SUM(EMPREGADO.salario * TRABALHA.horas / 160) AS valor_mensal_projeto`: Calcula o valor mensal a ser pago por projeto. O cálculo assume que o salário é mensal e que uma jornada de trabalho completa equivale a 160 horas por mês. Multiplicamos o salário do empregado pelo número de horas que ele dedica ao projeto, dividido por 160, para obter a contribuição mensal para o projeto. A função `SUM()` soma essas contribuições para obter o valor total a ser pago por cada projeto. O resultado é renomeado como `valor_mensal_projeto`.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta começa a selecionar dados da tabela `EMPREGADO`.

**JOIN Clauses**
   - `JOIN TRABALHA ON EMPREGADO.ssn = TRABALHA.essn`: Realiza uma junção entre a tabela `EMPREGADO` e `TRABALHA` onde o número de segurança social do empregado (`ssn`) corresponde ao número de segurança social (`essn`) na tabela `TRABALHA`. Isso permite combinar informações sobre os empregados com os projetos nos quais eles estão trabalhando.
   - `JOIN PROJETO ON TRABALHA.pno = PROJETO.pnumero`: Realiza uma junção entre a tabela `TRABALHA` e `PROJETO`, associando cada projeto ao número do projeto (`pnumero`).

**GROUP BY Clause**
   - `GROUP BY PROJETO.pjnome`: Agrupa os registros por nome do projeto (`pjnome`), garantindo que o valor total seja calculado para cada projeto individualmente.

# Atividade 4) – Subconsultas 

- Escreva subconsultas SQL para as questões:

a) Recupere nome (pnome e unome) de cada um dos empregados que tenham um dependente
cujo primeiro nome e sexo sejam o mesmo do empregado em questão.

In [43]:
cursor.execute("""SELECT pnome, unome
FROM EMPREGADO
WHERE EXISTS (
    SELECT 1
    FROM DEPENDENTE
    WHERE DEPENDENTE.essn = EMPREGADO.ssn
    AND DEPENDENTE.nomedep = EMPREGADO.pnome
    AND DEPENDENTE.sexodep = EMPREGADO.sexo
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

**SELECT Clause**
   - `pnome, unome`: Seleciona as colunas `pnome` (primeiro nome) e `unome` (sobrenome) da tabela `EMPREGADO`. O objetivo é retornar os nomes dos empregados que atendem às condições especificadas na subconsulta.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre os empregados, incluindo seus nomes, números de segurança social (`ssn`), e outros atributos.

**WHERE Clause**
   - `WHERE EXISTS`: Utiliza a cláusula `EXISTS` para verificar se uma determinada condição é verdadeira para pelo menos um registro retornado pela subconsulta. A consulta principal retorna um resultado apenas se a subconsulta retornar pelo menos um registro.
   
   **Subconsulta**:
   - **`SELECT 1`**: A subconsulta retorna o valor `1` para cada dependente que atende às condições especificadas. Não precisamos selecionar uma coluna específica, pois estamos apenas verificando a existência de registros que satisfaçam as condições.
   - **`FROM DEPENDENTE`**: A subconsulta obtém dados da tabela `DEPENDENTE`, que contém informações sobre os dependentes dos empregados, como o número de segurança social (`essn`), nome (`nomedep`), sexo (`sexodep`), e outros atributos.
   - **`WHERE DEPENDENTE.essn = EMPREGADO.ssn`**: Essa condição garante que a subconsulta apenas considere dependentes que estejam associados ao empregado atual da consulta principal, através do número de segurança social (`ssn`).
   - **`AND DEPENDENTE.nomedep = EMPREGADO.pnome`**: Verifica se o nome do dependente (`nomedep`) é igual ao primeiro nome do empregado (`pnome`). Esta condição é usada para encontrar dependentes que tenham o mesmo nome que o empregado.
   - **`AND DEPENDENTE.sexodep = EMPREGADO.sexo`**: Verifica se o sexo do dependente (`sexodep`) é igual ao sexo do empregado (`sexo`). Esta condição é usada para garantir que o dependente tenha o mesmo sexo que o empregado.

Nenhum dos dependentes tem o mesmo primeiro nome e sexo que os empregados associados. Por isso, a consulta SQL não retornou nenhum resultado.

b) Recupere os nomes dos empregados (pnome e unome) cujos salários são maiores que a
média dos salários dos empregados do departamento 5.

In [44]:
cursor.execute("""SELECT pnome, unome
FROM EMPREGADO
WHERE salario > (
    SELECT AVG(salario)
    FROM EMPREGADO
    WHERE dno = '5'
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Fabio', 'Will')
('Jennifer', 'Wallace')
('Ricardo', 'Nantes')
('Joaquim', 'Brito')
('Roberto', 'Silva')


**SELECT Clause**
   - `pnome, unome`: Seleciona as colunas `pnome` (primeiro nome) e `unome` (sobrenome) da tabela `EMPREGADO`. O objetivo é retornar os nomes dos empregados que têm salários maiores que a média dos salários dos empregados do departamento 5.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre os empregados, incluindo seus nomes, salários (`salario`), números de segurança social (`ssn`), e o número do departamento ao qual pertencem (`dno`).

**WHERE Clause**
   - `WHERE salario > (...)`: A condição `WHERE` compara o salário de cada empregado com o resultado da subconsulta. Apenas os empregados cujo salário for maior que a média calculada pela subconsulta serão retornados.

   **Subconsulta**:
   - **`SELECT AVG(salario)`**: A subconsulta calcula a média dos salários dos empregados no departamento 5, usando a função agregada `AVG()` para obter a média.
   - **`FROM EMPREGADO`**: Especifica que a subconsulta vai obter os dados da tabela `EMPREGADO`.
   - **`WHERE dno = '5'`**: Filtra os empregados para considerar apenas aqueles que trabalham no departamento 5 (`dno = '5'`).

c) Retorne o número do seguro social (SSN) de todos os empregado que trabalham com a
mesma combinação (projeto, horas) em algum dos projetos em que o empregado ‘Fabio Will’
(SSN= 333445555) trabalhe.

In [50]:
cursor.execute("""SELECT DISTINCT essn
FROM TRABALHA
WHERE (pno, horas) IN (
    SELECT pno, horas
    FROM TRABALHA
    WHERE essn = '333445555'
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

**SELECT Clause**
   - `DISTINCT essn`: Seleciona a coluna `essn` (número do seguro social) da tabela `TRABALHA`. O uso de `DISTINCT` garante que cada SSN apareça apenas uma vez no resultado, mesmo que um empregado trabalhe em múltiplas combinações iguais de projeto e horas.

**FROM Clause**
   - `FROM TRABALHA`: Especifica que a consulta vai obter os dados da tabela `TRABALHA`, que contém informações sobre os projetos nos quais os empregados estão envolvidos, incluindo o número do seguro social (`essn`), o número do projeto (`pno`), e as horas dedicadas ao projeto (`horas`).

**WHERE Clause**
   - `WHERE (pno, horas) IN (...)`: Esta cláusula filtra os empregados para incluir apenas aqueles que trabalham na mesma combinação de projeto (`pno`) e horas (`horas`) que "Fabio Will" (SSN = 333445555).

   **Subconsulta**:
   - **`SELECT pno, horas`**: A subconsulta seleciona o número do projeto (`pno`) e as horas (`horas`) que "Fabio Will" dedicou a cada projeto.
   - **`FROM TRABALHA`**: A subconsulta obtém os dados da tabela `TRABALHA`.
   - **`WHERE essn = '333445555'`**: Filtra a subconsulta para considerar apenas as combinações de projeto e horas associadas ao empregado "Fabio Will" (SSN = 333445555).

Como "Fabio Will" não tem registros na tabela TRABALHA, não há como encontrar outros empregados que trabalhem nas mesmas combinações de projeto e horas que ele. É por isso que a consulta não retornou resultado.

d) Recupere os nomes de todos os empregados que não trabalham em nenhum projeto.

In [51]:
cursor.execute("""SELECT pnome, unome
FROM EMPREGADO
WHERE ssn NOT IN (
    SELECT essn
    FROM TRABALHA
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Fabio', 'Will')
('Roberto', 'Silva')


**SELECT Clause**
   - `pnome, unome`: Seleciona as colunas `pnome` (primeiro nome) e `unome` (sobrenome) da tabela `EMPREGADO`. O objetivo é retornar os nomes dos empregados que não têm nenhum registro na tabela `TRABALHA`, ou seja, que não estão associados a nenhum projeto.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre todos os empregados, incluindo seus nomes e números de segurança social (`ssn`).

**WHERE Clause**
   - `WHERE ssn NOT IN (...)`: A cláusula `WHERE` filtra os empregados, retornando apenas aqueles cujo número de segurança social (`ssn`) não aparece na subconsulta. Isso significa que esses empregados não têm registros na tabela `TRABALHA` e, portanto, não estão associados a nenhum projeto.

   **Subconsulta**:
   - **`SELECT essn`**: A subconsulta seleciona a coluna `essn` da tabela `TRABALHA`, que contém os números de segurança social dos empregados que estão associados a projetos.
   - **`FROM TRABALHA`**: Especifica que a subconsulta vai obter os dados da tabela `TRABALHA`, onde cada registro representa um empregado trabalhando em um projeto específico.

e) Recupere o nome de empregados que não tenham dependentes.

In [52]:
cursor.execute("""SELECT unome
FROM EMPREGADO
WHERE ssn IN (
    SELECT gerssn
    FROM DEPARTAMENTO
)
AND ssn NOT IN (
    SELECT essn
    FROM DEPENDENTE
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Brito',)


**SELECT Clause**
   - `unome`: Seleciona a coluna `unome` (último nome) da tabela `EMPREGADO`. O objetivo é retornar os sobrenomes dos gerentes de departamento que não têm dependentes.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre todos os empregados, incluindo seus nomes, números de segurança social (`ssn`), e outros detalhes.

**WHERE Clause**
   - `WHERE ssn IN (...)`: Esta condição filtra os empregados para incluir apenas aqueles cujo número de segurança social (`ssn`) aparece na subconsulta, que identifica os gerentes de departamento.
   
   **Primeira Subconsulta**:
   - **`SELECT gerssn`**: A subconsulta seleciona a coluna `gerssn` da tabela `DEPARTAMENTO`, que contém os números de segurança social dos gerentes de departamento.
   - **`FROM DEPARTAMENTO`**: Especifica que a subconsulta vai obter os dados da tabela `DEPARTAMENTO`, que contém informações sobre os departamentos, incluindo os gerentes.

   - `AND ssn NOT IN (...)`: Esta condição adicional filtra para incluir apenas os gerentes de departamento cujo número de segurança social (`ssn`) não aparece na subconsulta, que identifica os empregados com dependentes.
   
   **Segunda Subconsulta**:
   - **`SELECT essn`**: A subconsulta seleciona a coluna `essn` da tabela `DEPENDENTE`, que contém os números de segurança social dos empregados que têm dependentes.
   - **`FROM DEPENDENTE`**: Especifica que a subconsulta vai obter os dados da tabela `DEPENDENTE`, onde cada registro representa um dependente associado a um empregado.

f) Liste o último nome de todos os gerentes de departamento que não tenham dependentes.

In [54]:
cursor.execute("""SELECT unome
FROM EMPREGADO
WHERE ssn IN (
    SELECT gerssn
    FROM DEPARTAMENTO
)
AND ssn NOT IN (
    SELECT essn
    FROM DEPENDENTE
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Brito',)


**SELECT Clause**
   - `unome`: Seleciona a coluna `unome` (último nome) da tabela `EMPREGADO`. O objetivo é retornar os sobrenomes dos gerentes de departamento que não têm dependentes.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre todos os empregados, incluindo seus nomes, números de segurança social (`ssn`), e outros detalhes.

**WHERE Clause**
   - `WHERE ssn IN (...)`: Filtra os empregados para incluir apenas aqueles cujo número de segurança social (`ssn`) aparece na subconsulta que identifica os gerentes de departamento.
   
   **Primeira Subconsulta**:
   - **`SELECT gerssn`**: A subconsulta seleciona a coluna `gerssn` da tabela `DEPARTAMENTO`, que contém os números de segurança social dos gerentes de departamento.
   - **`FROM DEPARTAMENTO`**: Especifica que a subconsulta vai obter os dados da tabela `DEPARTAMENTO`, que contém informações sobre os departamentos, incluindo os gerentes.

   - `AND ssn NOT IN (...)`: Esta condição adicional filtra para incluir apenas os gerentes de departamento cujo número de segurança social (`ssn`) **não** aparece na subconsulta que identifica os empregados com dependentes.
   
   **Segunda Subconsulta**:
   - **`SELECT essn`**: A subconsulta seleciona a coluna `essn` da tabela `DEPENDENTE`, que contém os números de segurança social dos empregados que têm dependentes.
   - **`FROM DEPENDENTE`**: Especifica que a subconsulta vai obter os dados da tabela `DEPENDENTE`, onde cada registro representa um dependente associado a um empregado.

g) Liste os nomes dos gerentes que tenham, pelo menos, um dependente.

In [53]:
cursor.execute("""SELECT pnome, unome
FROM EMPREGADO
WHERE ssn IN (
    SELECT gerssn
    FROM DEPARTAMENTO
)
AND ssn IN (
    SELECT essn
    FROM DEPENDENTE
);
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Fabio', 'Will')
('Jennifer', 'Wallace')


**SELECT Clause**
   - `pnome, unome`: Seleciona as colunas `pnome` (primeiro nome) e `unome` (sobrenome) da tabela `EMPREGADO`. O objetivo é retornar os nomes dos gerentes de departamento que têm pelo menos um dependente.

**FROM Clause**
   - `FROM EMPREGADO`: Especifica que a consulta vai obter os dados da tabela `EMPREGADO`, que contém informações sobre todos os empregados, incluindo seus nomes, números de segurança social (`ssn`), e outros detalhes.

**WHERE Clause**
   - `WHERE ssn IN (...)`: Filtra os empregados para incluir apenas aqueles cujo número de segurança social (`ssn`) aparece na subconsulta que identifica os gerentes de departamento.
   
   **Primeira Subconsulta**:
   - **`SELECT gerssn`**: A subconsulta seleciona a coluna `gerssn` da tabela `DEPARTAMENTO`, que contém os números de segurança social dos gerentes de departamento.
   - **`FROM DEPARTAMENTO`**: Especifica que a subconsulta vai obter os dados da tabela `DEPARTAMENTO`, que contém informações sobre os departamentos, incluindo os gerentes.

   - `AND ssn IN (...)`: Esta condição adicional filtra para incluir apenas os gerentes de departamento cujo número de segurança social (`ssn`) aparece na subconsulta que identifica os empregados com dependentes.
   
   **Segunda Subconsulta**:
   - **`SELECT essn`**: A subconsulta seleciona a coluna `essn` da tabela `DEPENDENTE`, que contém os números de segurança social dos empregados que têm dependentes.
   - **`FROM DEPENDENTE`**: Especifica que a subconsulta vai obter os dados da tabela `DEPENDENTE`, onde cada registro representa um dependente associado a um empregado.

# Atividade 5) Índices 

- Escreva comandos SQL para responder as seguintes questões:

a) Crie a tabela tempEmpregado contendo dos dados dos atributos ssn, pnome,sexo, endereco
e datanasc da tabela empregado. Realizar a copia dos dados durante o processo de criação da
nova tabela.

In [97]:
cursor.execute("""CREATE TABLE tempEmpregado AS
SELECT ssn, pnome, sexo, endereco, datanasc
FROM EMPREGADO;
""")

conn.commit()

- **`CREATE TABLE tempEmpregado AS`**: Cria uma nova tabela chamada `tempEmpregado`. A cláusula `AS` indica que a nova tabela será criada com base nos resultados de uma consulta SQL.
- **`SELECT ssn, pnome, sexo, endereco, datanasc`**: Seleciona as colunas `ssn`, `pnome`, `sexo`, `endereco`, e `datanasc` da tabela `EMPREGADO`.
- **`FROM EMPREGADO`**: Especifica que os dados para a nova tabela `tempEmpregado` serão obtidos da tabela `EMPREGADO`.

b) Utilize o comando EXPLAIN ANALYSE e anote o tempo de execução da consulta SELECT *
FROM TEMPEMPREGADO.

In [98]:
cursor.execute("""EXPLAIN ANALYZE SELECT * FROM tempEmpregado;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Seq Scan on tempempregado  (cost=0.00..12.90 rows=290 width=248) (actual time=0.014..0.015 rows=9 loops=1)',)
('Planning Time: 1.119 ms',)
('Execution Time: 0.039 ms',)


**Seq Scan on tempempregado**:
   - **Custo estimado (`cost`)**: O PostgreSQL estimou que o custo da execução dessa consulta seria entre `0.00` e `12.90`.
   - **Linhas estimadas (`rows`)**: O planejador estimou que seriam retornadas `290` linhas.
   - **Tempo real de execução (`actual time`)**: A consulta realmente levou entre `0.014` e `0.015` milissegundos para executar essa etapa, retornando `9` linhas.

**Planning Time**:
   - O tempo necessário para o PostgreSQL planejar a execução da consulta foi de `1.119 ms`.

**Execution Time**:
   - O tempo total de execução da consulta foi de `0.039 ms`.

- **`EXPLAIN ANALYZE`**: Este comando é usado para analisar a execução de uma consulta SQL, mostrando o plano de execução que o PostgreSQL planeja usar e o tempo necessário para executá-la. O `ANALYZE` realmente executa a consulta, ao contrário de apenas mostrar o plano, permitindo ver o tempo real de execução.
- **`SELECT * FROM tempEmpregado`**: Esta é a consulta que está sendo analisada, que simplesmente seleciona todos os registros e colunas da tabela `tempEmpregado`.

c) Crie um índice com o campo ssn que não permite dados duplicados para a tabela
tempEmpregado, utilizando o algoritmo de ordenação BTREE.

In [99]:
cursor.execute("""CREATE UNIQUE INDEX idx_tempempregado_ssn
ON tempEmpregado (ssn);
""")
conn.commit()

No PostgreSQL, o algoritmo `BTREE` é o padrão para a criação de índices.

- **`CREATE UNIQUE INDEX idx_tempempregado_ssn`**: Cria um índice exclusivo chamado `idx_tempempregado_ssn` na tabela `tempEmpregado`.
- **`ON tempEmpregado (ssn)`**: Especifica que o índice será criado na coluna `ssn` da tabela `tempEmpregado`.

d) Crie um índice com o campo pnome para a tabela tempEmpregado, em ordem decrescente
e que utilize o algoritmo de ordenação HASH.

In [100]:
cursor.execute("""CREATE INDEX idx_tempempregado_pnome_desc
ON tempEmpregado (pnome DESC);
""")

conn.commit()

- **`CREATE INDEX idx_tempempregado_pnome_desc`**: Cria um índice chamado `idx_tempempregado_pnome_desc` na tabela `tempEmpregado`.
- **`ON tempEmpregado (pnome DESC)`**: Especifica que o índice será criado na coluna `pnome` em ordem decrescente (`DESC`).

e) Utilize o comando EXPLAIN ANALYSE, anote e compare com o tempo de execução da
consulta SELECT * FROM TEMPEMPREGADO. O que achou do resultado?

In [101]:
cursor.execute("""EXPLAIN ANALYZE SELECT * FROM tempEmpregado;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Seq Scan on tempempregado  (cost=0.00..1.09 rows=9 width=248) (actual time=0.026..0.029 rows=9 loops=1)',)
('Planning Time: 5.070 ms',)
('Execution Time: 0.055 ms',)


**Seq Scan (Varredura Sequencial)**:
   - Em ambos os casos, o PostgreSQL usou uma varredura sequencial (`Seq Scan`) para acessar os dados, o que é esperado, já que a consulta `SELECT *` não se beneficia diretamente do índice criado na coluna `pnome`.
   - **Custo estimado**: O custo estimado diminuiu após a criação do índice (`12.90` para `1.09`), indicando que o planejador de consultas considerou o índice na estimativa de custos, mesmo que a execução real não tenha se beneficiado diretamente dele.
   - **Tempo real de execução**: Houve um pequeno aumento no tempo de execução (`0.039 ms` para `0.055 ms`), provavelmente devido ao overhead de manutenção do índice durante a consulta, mesmo que ele não tenha sido diretamente utilizado.

**Planning Time**:
   - O tempo de planejamento (`Planning Time`) aumentou significativamente após a criação do índice (`1.119 ms` para `5.070 ms`). Isso pode ocorrer porque o planejador de consultas agora precisa considerar o índice ao determinar o melhor plano de execução, mesmo que o índice não seja utilizado diretamente na execução da consulta.

**Execution Time**:
   - O tempo total de execução (`Execution Time`) aumentou ligeiramente após a criação do índice (`0.039 ms` para `0.055 ms`). Este pequeno aumento pode ser atribuído ao fato de que, embora a varredura sequencial continue a ser utilizada, o sistema agora também leva em conta o índice, o que pode introduzir algum overhead.

### Conclusão

- **Índice não utilizado diretamente**: Neste caso específico, a criação do índice na coluna `pnome` não trouxe benefícios diretos para a consulta `SELECT * FROM tempEmpregado`, já que a consulta não estava filtrando ou ordenando os dados com base na coluna indexada.
- **Overhead do índice**: A criação do índice introduziu um pequeno overhead, aumentando o tempo de planejamento e execução. Isso é comum quando índices são criados em colunas que não são diretamente utilizadas nas consultas.
- **Uso do índice em outras consultas**: O índice poderia ser mais útil em outras consultas que filtram ou ordenam com base na coluna `pnome`. Se você tivesse uma consulta como `SELECT * FROM tempEmpregado WHERE pnome = 'Joao'`, o índice poderia acelerar significativamente a execução.

f) Remova os índices criados nos itens c e d.

In [102]:
cursor.execute("""DROP INDEX IF EXISTS idx_tempempregado_ssn;
""")

cursor.execute("""DROP INDEX IF EXISTS idx_tempempregado_pnome_desc;
""")
conn.commit()

g) Remova a tabela temporária criada no item a

In [103]:
cursor.execute("""DROP TABLE IF EXISTS tempEmpregado;
""")
conn.commit()

# Atividade 6) Visões 

- Escreva comandos SQL para responder as seguintes questões

a) Crie a visão chamada TRABALHA_EM que deverá conter os campos pnome e unome da
tabela empregado, o campo pjnome da tabela projeto e o campo horas da tabela trabalha

In [116]:
cursor.execute("""CREATE VIEW TRABALHA_EM AS
SELECT EMPREGADO.pnome, EMPREGADO.unome, PROJETO.pjnome, TRABALHA.horas
FROM EMPREGADO
JOIN TRABALHA ON EMPREGADO.ssn = TRABALHA.essn
JOIN PROJETO ON TRABALHA.pno = PROJETO.pnumero;
""")

conn.commit()

- **`CREATE VIEW TRABALHA_EM AS`**: Cria uma nova visão chamada `TRABALHA_EM`. Uma visão é uma tabela virtual que é baseada no resultado de uma consulta SQL.
- **`SELECT EMPREGADO.pnome, EMPREGADO.unome, PROJETO.pjnome, TRABALHA.horas`**: Seleciona os campos `pnome` e `unome` da tabela `EMPREGADO`, o campo `pjnome` da tabela `PROJETO`, e o campo `horas` da tabela `TRABALHA`.
- **`FROM EMPREGADO`**: Especifica que a consulta começa selecionando dados da tabela `EMPREGADO`.
- **`JOIN TRABALHA ON EMPREGADO.ssn = TRABALHA.essn`**: Faz uma junção entre a tabela `EMPREGADO` e `TRABALHA` com base na correspondência entre o número de segurança social do empregado (`ssn`) na tabela `EMPREGADO` e o número de segurança social (`essn`) na tabela `TRABALHA`.
- **`JOIN PROJETO ON TRABALHA.pno = PROJETO.pnumero`**: Faz uma junção entre a tabela `TRABALHA` e `PROJETO` com base na correspondência entre o número do projeto (`pno`) na tabela `TRABALHA` e o número do projeto (`pnumero`) na tabela `PROJETO`.

b) Crie uma consulta SQL na visão implementada no item a que retorne o último e o primeiro
nome de todos os empregados que trabalham no ‘ProdutoX’.

In [117]:
cursor.execute("""SELECT unome, pnome
FROM TRABALHA_EM
WHERE pjnome = 'ProdutoX';
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Souza', 'Joao')
('Pereira', 'Jussara')


- **`SELECT unome, pnome`**: Seleciona os campos `unome` (último nome) e `pnome` (primeiro nome) dos empregados da visão `TRABALHA_EM`.
- **`FROM TRABALHA_EM`**: Especifica que a consulta vai buscar os dados da visão `TRABALHA_EM`, que contém a junção das tabelas `EMPREGADO`, `TRABALHA` e `PROJETO`.
- **`WHERE pjnome = 'ProdutoX'`**: Filtra os resultados para incluir apenas os empregados que estão trabalhando no projeto com o nome 'ProdutoX'.

c) Exclua a visão criada no item a.

In [118]:
cursor.execute("""DROP VIEW IF EXISTS TRABALHA_EM;
""")

conn.commit()

d) Crie uma visão chamada DEPTO_INFO que deverá conter os campos dnome da tabela
departamento, e o total de empregados e somatório dos salários dos empregados da tabela
empregado por departamento

In [119]:
cursor.execute("""CREATE VIEW DEPTO_INFO AS
SELECT DEPARTAMENTO.dnome, 
       COUNT(EMPREGADO.ssn) AS total_empregados, 
       SUM(EMPREGADO.salario) AS soma_salarios
FROM DEPARTAMENTO
LEFT JOIN EMPREGADO ON DEPARTAMENTO.dnumero = EMPREGADO.dno
GROUP BY DEPARTAMENTO.dnome;
""")

conn.commit()

- **`CREATE VIEW DEPTO_INFO AS`**: Cria uma nova visão chamada `DEPTO_INFO`.
- **`SELECT DEPARTAMENTO.dnome`**: Seleciona o nome do departamento (`dnome`) da tabela `DEPARTAMENTO`.
- **`COUNT(EMPREGADO.ssn) AS total_empregados`**: Conta o número de empregados em cada departamento, identificando-os pelo número de segurança social (`ssn`). O resultado é renomeado como `total_empregados`.
- **`SUM(EMPREGADO.salario) AS soma_salarios`**: Calcula o somatório dos salários dos empregados em cada departamento. O resultado é renomeado como `soma_salarios`.
- **`FROM DEPARTAMENTO`**: Especifica que a consulta vai buscar dados da tabela `DEPARTAMENTO`.
- **`LEFT JOIN EMPREGADO ON DEPARTAMENTO.dnumero = EMPREGADO.dno`**: Realiza uma junção à esquerda (`LEFT JOIN`) entre as tabelas `DEPARTAMENTO` e `EMPREGADO` com base na correspondência entre o número do departamento (`dnumero`) na tabela `DEPARTAMENTO` e o número do departamento (`dno`) na tabela `EMPREGADO`. Isso garante que todos os departamentos sejam incluídos na visão, mesmo que não tenham empregados associados.
- **`GROUP BY DEPARTAMENTO.dnome`**: Agrupa os resultados por nome do departamento, de forma que a contagem de empregados e a soma dos salários sejam calculadas para cada departamento individualmente.

e) Crie uma consulta SQL na visão implementada no item c que retorne a lista de informações
por departamentos ordenados pelo somatório dos salários.

In [120]:
cursor.execute("""SELECT dnome, total_empregados, soma_salarios
FROM DEPTO_INFO
ORDER BY soma_salarios DESC;
""")

# Obter resultados
results = cursor.fetchall()
if results:
    for result in results:
        print(result)

('Pesquisa', 4, 133000.0)
('Sede Administrativa', 2, 113000.0)
('Administracao', 3, 93000.0)


- **`SELECT dnome, total_empregados, soma_salarios`**: Seleciona as colunas `dnome` (nome do departamento), `total_empregados` (total de empregados), e `soma_salarios` (somatório dos salários) da visão `DEPTO_INFO`.
- **`FROM DEPTO_INFO`**: Especifica que a consulta vai buscar os dados da visão `DEPTO_INFO`.
- **`ORDER BY soma_salarios DESC`**: Ordena os resultados pelo somatório dos salários (`soma_salarios`) em ordem decrescente (`DESC`), de modo que os departamentos com os maiores somatórios de salários apareçam primeiro.

f) Exclua as visões criadas nos itens a.

In [121]:
cursor.execute("""DROP VIEW IF EXISTS DEPTO_INFO;
""")

conn.commit()